File: paimon-core/src/main/java/org/apache/paimon/rest/requests/TableSchema.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.paimon.rest.requests;

import org.apache.paimon.schema.Schema;
import org.apache.paimon.types.DataField;

import org.apache.paimon.shade.jackson2.com.fasterxml.jackson.annotation.JsonCreator;
import org.apache.paimon.shade.jackson2.com.fasterxml.jackson.annotation.JsonGetter;
import org.apache.paimon.shade.jackson2.com.fasterxml.jackson.annotation.JsonProperty;

import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * Wrap the {@link Schema} class to support RESTCatalog. Define a class as: 1. This class in rest
 * catalog is easy to maintain. 2. It's easy to manage rest API fields.
 */
public class TableSchema {
    private static final String FIELD_FILED_NAME = "fields";
    private static final String FIELD_PARTITION_KEYS_NAME = "partitionKeys";
    private static final String FIELD_PRIMARY_KEYS_NAME = "primaryKeys";
    private static final String FIELD_OPTIONS_NAME = "options";
    private static final String FIELD_COMMENT_NAME = "comment";

    @JsonProperty(FIELD_FILED_NAME)
    private final List<DataField> fields;

    @JsonProperty(FIELD_PARTITION_KEYS_NAME)
    private final List<String> partitionKeys;

    @JsonProperty(FIELD_PRIMARY_KEYS_NAME)
    private final List<String> primaryKeys;

    @JsonProperty(FIELD_OPTIONS_NAME)
    private final Map<String, String> options;

    @JsonProperty(FIELD_COMMENT_NAME)
    private final String comment;

    @JsonCreator
    public TableSchema(
            @JsonProperty(FIELD_FILED_NAME) List<DataField> fields,
            @JsonProperty(FIELD_PARTITION_KEYS_NAME) List<String> partitionKeys,
            @JsonProperty(FIELD_PRIMARY_KEYS_NAME) List<String> primaryKeys,
            @JsonProperty(FIELD_OPTIONS_NAME) Map<String, String> options,
            @JsonProperty(FIELD_COMMENT_NAME) String comment) {
        this.fields = fields;
        this.partitionKeys = partitionKeys;
        this.primaryKeys = primaryKeys;
        this.options = options;
        this.comment = comment;
    }

    public TableSchema(Schema schema) {
        this.fields = schema.fields();
        this.partitionKeys = schema.partitionKeys();
        this.primaryKeys = schema.primaryKeys();
        this.options = schema.options();
        this.comment = schema.comment();
    }

    @JsonGetter(FIELD_FILED_NAME)
    public List<DataField> getFields() {
        return fields;
    }

    @JsonGetter(FIELD_PARTITION_KEYS_NAME)
    public List<String> getPartitionKeys() {
        return partitionKeys;
    }

    @JsonGetter(FIELD_PRIMARY_KEYS_NAME)
    public List<String> getPrimaryKeys() {
        return primaryKeys;
    }

    @JsonGetter(FIELD_OPTIONS_NAME)
    public Map<String, String> getOptions() {
        return options;
    }

    @JsonGetter(FIELD_COMMENT_NAME)
    public String getComment() {
        return comment;
    }

    @Override
    public boolean equals(Object o) {
        if (o == null || getClass() != o.getClass()) {
            return false;
        } else {
            TableSchema that = (TableSchema) o;
            return Objects.equals(fields, that.fields)
                    && Objects.equals(partitionKeys, that.partitionKeys)
                    && Objects.equals(primaryKeys, that.primaryKeys)
                    && Objects.equals(options, that.options)
                    && Objects.equals(comment, that.comment);
        }
    }

    @Override
    public int hashCode() {
        return Objects.hash(fields, partitionKeys, primaryKeys, options, comment);
    }
}

File: paimon-python/pypaimon/api/table_schema.py
"""
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""

import time
from typing import List, Dict, Optional

from pypaimon.api.schema import Schema
from pypaimon.api.data_types import DataField


class TableSchema:

    def __init__(self, id: int, fields: List[DataField], highest_field_id: int,
                 partition_keys: List[str], primary_keys: List[str], options: Dict[str, str],
                 comment: Optional[str] = None, time_millis: Optional[int] = None):
        self.id = id
        self.fields = fields
        self.highest_field_id = highest_field_id
        self.partition_keys = partition_keys or []
        self.primary_keys = primary_keys or []
        self.options = options or {}
        self.comment = comment
        self.time_millis = time_millis if time_millis is not None else int(time.time() * 1000)

    def to_schema(self) -> Schema:
        # pa_schema = schema_util.convert_data_fields_to_pa_schema(self.fields)
        return Schema(
            fields=self.fields,
            partition_keys=self.partition_keys,
            primary_keys=self.primary_keys,
            options=self.options,
            comment=self.comment
        )

    @staticmethod
    def create(schema_id: int, schema: Schema) -> "TableSchema":
        fields: List[DataField] = schema.fields

        partition_keys: List[str] = schema.partition_keys

        primary_keys: List[str] = schema.primary_keys

        options: Dict[str, str] = schema.options

        highest_field_id: int = None

        return TableSchema(
            schema_id,
            fields,
            highest_field_id,
            partition_keys,
            primary_keys,
            options,
            schema.comment,
            int(time.time())
        )

    def copy(self, new_options: Optional[Dict[str, str]] = None) -> "TableSchema":
        return TableSchema(
            id=self.id,
            fields=self.fields,
            highest_field_id=self.highest_field_id,
            partition_keys=self.partition_keys,
            primary_keys=self.primary_keys,
            options=new_options,
            comment=self.comment,
            time_millis=self.time_millis
        )
